File: Delphi/ClipperCore.pas
unit ClipperCore;

(*******************************************************************************
* Author    :  Angus Johnson                                                   *
* Version   :  10.0 (beta)                                                     *
* Date      :  14 January 2019                                                 *
* Website   :  http://www.angusj.com                                           *
* Copyright :  Angus Johnson 2010-2019                                         *
* Purpose   :  Core Clipper Library module                                     *
*              Contains structures and functions used throughout the library   *
* License   :  http://www.boost.org/LICENSE_1_0.txt                            *
*******************************************************************************)

{$IFDEF FPC}
  {$DEFINE INLINING}
{$ELSE}
  {$IF CompilerVersion < 14}
    Requires Delphi version 6 or above.
  {$IFEND}
  {$IF CompilerVersion >= 18}         //Delphi 2007
    //While Inlining has been supported since D2005, both D2005 and D2006
    //have an Inline codegen bug (QC41166) so ignore Inline until D2007.
    {$DEFINE INLINING}
    {$IF CompilerVersion >= 25.0}     //Delphi XE4+
      {$LEGACYIFEND ON}
    {$IFEND}
  {$IFEND}
{$ENDIF}

{$IFDEF DEBUG}
  {$UNDEF INLINING}
{$ENDIF}

interface

uses
  Classes, SysUtils, Math;

type
  TPoint64 = record X, Y: Int64; end;
  TPointD = record X, Y: double; end;

  //TPath: a simple data structure to represent a series of vertices, whether
  //open (poly-line) or closed (polygon). A path may be simple or complex (self
  //intersecting). For simple polygons, path orientation (whether clockwise or
  //counter-clockwise) is generally used to differentiate outer paths from inner
  //paths (holes). For complex polygons (and also for overlapping polygons),
  //explicit 'filling rules' (see below) are used to indicate regions that are
  //inside (filled) and regions that are outside (unfilled) a specific polygon.
  TPath = array of TPoint64;
  TPaths = array of TPath;
  TArrayOfPaths = array of TPaths;

  TPathD = array of TPointD;
  TPathsD = array of TPathD;

  TRect64 = {$IFDEF UNICODE}record{$ELSE}object{$ENDIF}
  private
    function GetWidth: Int64; {$IFDEF INLINING} inline; {$ENDIF}
    function GetHeight: Int64; {$IFDEF INLINING} inline; {$ENDIF}
    function GetIsEmpty: Boolean; {$IFDEF INLINING} inline; {$ENDIF}
  public
    Left   : Int64;
    Top    : Int64;
    Right  : Int64;
    Bottom : Int64;
    property Width: Int64 read GetWidth;
    property Height: Int64 read GetHeight;
    property IsEmpty: Boolean read GetIsEmpty;
  end;

  TRectD = {$ifdef UNICODE}record{$else}object{$endif}
  private
    function GetWidth: double; {$IFDEF INLINING} inline; {$ENDIF}
    function GetHeight: double; {$IFDEF INLINING} inline; {$ENDIF}
    function GetIsEmpty: Boolean; {$IFDEF INLINING} inline; {$ENDIF}
  public
    Left   : double;
    Top    : double;
    Right  : double;
    Bottom : double;
    property Width: double read GetWidth;
    property Height: double read GetHeight;
    property IsEmpty: Boolean read GetIsEmpty;
  end;

  TClipType = (ctNone, ctIntersection, ctUnion, ctDifference, ctXor);
  TPathType = (ptSubject, ptClip);
  //By far the most widely used winding rules for polygon filling are EvenOdd
  //and NonZero (see GDI, GDI+, XLib, OpenGL, Cairo, AGG, Quartz, SVG, Gr32).
  //https://www.w3.org/TR/SVG/painting.html
  TFillRule = (frEvenOdd, frNonZero, frPositive, frNegative);

  TPointInPolygonResult = (pipInside, pipOutside, pipOn);

  EClipperLibException = class(Exception);

function PointsEqual(const p1, p2: TPoint64): Boolean;
  {$IFDEF INLINING} inline; {$ENDIF}
function Point64(const X, Y: Int64): TPoint64; overload;
  {$IFDEF INLINING} inline; {$ENDIF}
function Point64(const X, Y: Double): TPoint64; overload;
  {$IFDEF INLINING} inline; {$ENDIF}
function PointD(const X, Y: Double): TPointD;
  {$IFDEF INLINING} inline; {$ENDIF}
function Rect64(const left, top, right, bottom: Int64): TRect64;
function RectD(const left, top, right, bottom: double): TRectD;
function GetBounds(const paths: TArrayOfPaths): TRect64; overload;
function GetBounds(const paths: TPaths): TRect64; overload;
function GetBounds(const paths: TPathsD): TRectD; overload;

function InflateRect(const rec: TRect64; dx, dy: Int64): TRect64;
function UnionRects(const rec, rec2: TRect64): TRect64;

//Area: result is type double to avoid potential integer overflow
function Area(const path: TPath): Double;
function Orientation(const path: TPath): Boolean;
function PointInPolygon(const pt: TPoint64;
  const path: TPath): TPointInPolygonResult;
function CrossProduct(const pt1, pt2, pt3: TPoint64): double;

function ScalePaths(const paths: TPaths; sx, sy: double): TPaths;
function ScalePathsD(const paths: TPaths; sx, sy: double): TPathsD; overload;
function ScalePathsD(const paths: TPathsD; sx, sy: double): TPathsD; overload;
function OffsetPaths(const paths: TPaths; dx, dy: Int64): TPaths;
function OffsetPathsD(const paths: TPathsD; dx, dy: double): TPathsD;

function ReversePath(const path: TPath): TPath;
function ReversePaths(const paths: TPaths): TPaths;

procedure AppendPaths(var paths: TPaths; const extra: TPaths);

function ArrayOfPathsToPaths(const ap: TArrayOfPaths): TPaths;

const
  nullRect: TRect64 = (left:0; top: 0; right:0; Bottom: 0);
  nullRectD: TRectD = (left:0.0; top: 0.0; right:0.0; Bottom: 0.0);

implementation

//------------------------------------------------------------------------------
// TRect64 methods ...
//------------------------------------------------------------------------------

function TRect64.GetWidth: Int64;
begin
  result := right - left;
end;
//------------------------------------------------------------------------------

function TRect64.GetHeight: Int64;
begin
  result := bottom - top;
end;
//------------------------------------------------------------------------------

function TRect64.GetIsEmpty: Boolean;
begin
  result := (bottom <= top) or (right <= left);
end;

//------------------------------------------------------------------------------
// TRectD methods ...
//------------------------------------------------------------------------------

function TRectD.GetWidth: double;
begin
  result := right - left;
end;
//------------------------------------------------------------------------------

function TRectD.GetHeight: double;
begin
  result := bottom - top;
end;
//------------------------------------------------------------------------------

function TRectD.GetIsEmpty: Boolean;
begin
  result := (bottom > top) and (right > left);
end;


//------------------------------------------------------------------------------
// Miscellaneous Functions ...
//------------------------------------------------------------------------------

procedure RaiseError(const msg: string); {$IFDEF INLINING} inline; {$ENDIF}
begin
  raise EClipperLibException.Create(msg);
end;
//------------------------------------------------------------------------------

function ScalePaths(const paths: TPaths; sx, sy: double): TPaths;
var
  i,j,k,len: integer;
begin
  //strips duplicates
  setlength(result, length(paths));
  for i := 0 to high(paths) do
  begin
    len := length(paths[i]);
    setlength(result[i], len);
    if len = 0 then Continue;
    result[i][0].X := Round(paths[i][0].X * sx);
    result[i][0].Y := Round(paths[i][0].Y * sy);
    k := 1;
    for j := 1 to len -1 do
    begin
      result[i][k].X := Round(paths[i][j].X * sx);
      result[i][k].Y := Round(paths[i][j].Y * sy);
      if (result[i][k].X <> result[i][k-1].X) or
        (result[i][k].Y <> result[i][k-1].Y) then inc(k);
    end;
    if (k > 1) and (result[i][k-1].X = result[i][0].X) and
        (result[i][k-1].Y = result[i][0].Y) then dec(k);
    SetLength(result[i], k);
  end;
end;
//------------------------------------------------------------------------------

function ScalePathsD(const paths: TPaths; sx, sy: double): TPathsD;
var
  i,j: integer;
begin
  setlength(result, length(paths));
  for i := 0 to high(paths) do
  begin
    setlength(result[i], length(paths[i]));
    for j := 0 to high(paths[i]) do
    begin
      result[i][j].X := paths[i][j].X * sx;
      result[i][j].Y := paths[i][j].Y * sy;
    end;
  end;
end;
//------------------------------------------------------------------------------

function ScalePathsD(const paths: TPathsD; sx, sy: double): TPathsD;
var
  i,j: integer;
begin
  setlength(result, length(paths));
  for i := 0 to high(paths) do
  begin
    setlength(result[i], length(paths[i]));
    for j := 0 to high(paths[i]) do
    begin
      result[i][j].X := paths[i][j].X * sx;
      result[i][j].Y := paths[i][j].Y * sy;
    end;
  end;
end;
//------------------------------------------------------------------------------

function OffsetPaths(const paths: TPaths; dx, dy: Int64): TPaths;
var
  i,j: integer;
begin
  setlength(result, length(paths));
  for i := 0 to high(paths) do
  begin
    setlength(result[i], length(paths[i]));
    for j := 0 to high(paths[i]) do
    begin
      result[i][j].X := paths[i][j].X + dx;
      result[i][j].Y := paths[i][j].Y + dy;
    end;
  end;
end;
//------------------------------------------------------------------------------

function OffsetPathsD(const paths: TPathsD; dx, dy: double): TPathsD;
var
  i,j: integer;
begin
  setlength(result, length(paths));
  for i := 0 to high(paths) do
  begin
    setlength(result[i], length(paths[i]));
    for j := 0 to high(paths[i]) do
    begin
      result[i][j].X := paths[i][j].X + dx;
      result[i][j].Y := paths[i][j].Y + dy;
    end;
  end;
end;
//------------------------------------------------------------------------------

function ReversePath(const path: TPath): TPath;
var
  i, highI: Integer;
begin
  highI := high(path);
  SetLength(Result, highI +1);
  for i := 0 to highI do
    Result[i] := path[highI - i];
end;
//------------------------------------------------------------------------------

function ReversePaths(const paths: TPaths): TPaths;
var
  i, j, highJ: Integer;
begin
  i := length(paths);
  SetLength(Result, i);
  for i := 0 to i -1 do
  begin
    highJ := high(paths[i]);
    SetLength(Result[i], highJ+1);
    for j := 0 to highJ do
      Result[i][j] := paths[i][highJ - j];
  end;
end;
//------------------------------------------------------------------------------

procedure AppendPaths(var paths: TPaths; const extra: TPaths);
var
  i, len1, len2: Integer;
begin
  len1 := length(paths);
  len2 := length(extra);
  SetLength(paths, len1 + len2);
  for i := 0 to len2 -1 do
    paths[len1 + i] := extra[i];
end;
//------------------------------------------------------------------------------

function ArrayOfPathsToPaths(const ap: TArrayOfPaths): TPaths;
var
  i,j,k, len, cnt: integer;
begin
  cnt := 0;
  len := length(ap);
  for i := 0 to len -1 do
    inc(cnt, length(ap[i]));
  k := 0;
  setlength(result, cnt);
  for i := 0 to len -1 do
    for j := 0 to length(ap[i]) -1 do
    begin
      result[k] := ap[i][j];
      inc(k);
    end;
end;
//------------------------------------------------------------------------------

function PointsEqual(const p1, p2: TPoint64): Boolean;
  {$IFDEF INLINING} inline; {$ENDIF}
begin
  Result := (p1.X = p2.X) and (p1.Y = p2.Y);
end;
//------------------------------------------------------------------------------

function Point64(const X, Y: Int64): TPoint64;
begin
  Result.X := X;
  Result.Y := Y;
end;
//------------------------------------------------------------------------------

function Point64(const X, Y: Double): TPoint64;
begin
  Result.X := Round(X);
  Result.Y := Round(Y);
end;
//------------------------------------------------------------------------------

function PointD(const X, Y: Double): TPointD;
begin
  Result.X := X;
  Result.Y := Y;
end;
//------------------------------------------------------------------------------

function Rect64(const left, top, right, bottom: Int64): TRect64;
begin
  Result.Left := left;
  Result.Top := top;
  Result.Right := right;
  Result.Bottom := bottom;
end;
//------------------------------------------------------------------------------

function RectD(const left, top, right, bottom: double): TRectD;
begin
  Result.Left := left;
  Result.Top := top;
  Result.Right := right;
  Result.Bottom := bottom;
end;
//------------------------------------------------------------------------------

function GetBounds(const paths: TArrayOfPaths): TRect64; overload;
var
  i,j,k: Integer;
begin
  Result := Rect64(High(Int64), High(Int64), Low(Int64), Low(Int64));
  for i := 0 to High(paths) do
    for j := 0 to High(paths[i]) do
      for k := 0 to High(paths[i][j]) do
      begin
        if paths[i][j][k].X < Result.Left then Result.Left := paths[i][j][k].X;
        if paths[i][j][k].X > Result.Right then Result.Right := paths[i][j][k].X;
        if paths[i][j][k].Y < Result.Top then Result.Top := paths[i][j][k].Y;
        if paths[i][j][k].Y > Result.Bottom then Result.Bottom := paths[i][j][k].Y;
      end;
  if Result.Left > Result.Right then Result := nullRect;
end;
//------------------------------------------------------------------------------

function GetBounds(const paths: TPaths): TRect64;
var
  i,j: Integer;
begin
  Result := Rect64(High(Int64), High(Int64), Low(Int64), Low(Int64));
  for i := 0 to High(paths) do
    for j := 0 to High(paths[i]) do
    begin
      if paths[i][j].X < Result.Left then Result.Left := paths[i][j].X;
      if paths[i][j].X > Result.Right then Result.Right := paths[i][j].X;
      if paths[i][j].Y < Result.Top then Result.Top := paths[i][j].Y;
      if paths[i][j].Y > Result.Bottom then Result.Bottom := paths[i][j].Y;
    end;
  if Result.Left > Result.Right then Result := nullRect;
end;
//------------------------------------------------------------------------------

function GetBounds(const paths: TPathsD): TRectD;
var
  i,j: Integer;
begin
  Result := RectD(MaxDouble, MaxDouble, -MaxDouble, -MaxDouble);
  for i := 0 to High(paths) do
    for j := 0 to High(paths[i]) do
    begin
      if paths[i][j].X < Result.Left then Result.Left := paths[i][j].X;
      if paths[i][j].X > Result.Right then Result.Right := paths[i][j].X;
      if paths[i][j].Y < Result.Top then Result.Top := paths[i][j].Y;
      if paths[i][j].Y > Result.Bottom then Result.Bottom := paths[i][j].Y;
    end;
  result.Left := Floor(Result.Left);
  result.Top := Floor(Result.Top);
  result.Right := Ceil(Result.Left);
  result.Bottom := Ceil(Result.Top);
  if Result.Left > Result.Right then Result := nullRectD;
end;
//------------------------------------------------------------------------------

function InflateRect(const rec: TRect64; dx, dy: Int64): TRect64;
begin
  result := rec;
  dec(result.Left, dx);
  inc(result.Right, dx);
  dec(result.Top, dy);
  inc(result.Bottom, dy);
end;
//------------------------------------------------------------------------------

function UnionRects(const rec, rec2: TRect64): TRect64;
begin
  if rec.IsEmpty then result := rec2
  else if rec2.IsEmpty then result := rec
  else
  begin
    result.Left := min(rec.Left, rec2.Left);
    result.Right := max(rec.Right, rec2.Right);
    result.Top := min(rec.Top, rec2.Top);
    result.Bottom := max(rec.Bottom, rec2.Bottom);
  end;
end;
//------------------------------------------------------------------------------

function Area(const path: TPath): Double;
var
  i, j, highI: Integer;
  d: Double;
begin
  Result := 0.0;
  highI := High(path);
  if (highI < 2) then Exit;
  j := highI;
  for i := 0 to highI do
  begin
    d := (path[j].X + path[i].X);
    Result := Result + d * (path[j].Y - path[i].Y);
    j := i;
  end;
  Result := -Result * 0.5;
end;
//------------------------------------------------------------------------------

function Orientation(const path: TPath): Boolean;
begin
  Result := Area(path) >= 0;
end;
//------------------------------------------------------------------------------

function PointInPolygon(const pt: TPoint64;
  const path: TPath): TPointInPolygonResult;
var
  i, val, cnt: Integer;
  d, d2, d3: Double; //using doubles to avoid possible integer overflow
  ip, ipNext: TPoint64;
begin
  cnt := Length(path);
  if cnt < 3 then
  begin
    result := pipOutside;
    Exit;
  end;
  ip := path[0];
  Result := pipOn;
  val := 0;
  for i := 1 to cnt do
  begin
    if i < cnt then ipNext := path[i]
    else ipNext := path[0];

    if (ipNext.Y = pt.Y) then
    begin
      if (ipNext.X = pt.X) or ((ip.Y = pt.Y) and
        ((ipNext.X > pt.X) = (ip.X < pt.X))) then Exit;
    end;

    if ((ip.Y < pt.Y) <> (ipNext.Y < pt.Y)) then
    begin
      if (ip.X >= pt.X) then
      begin
        if (ipNext.X > pt.X) then val := 1 - val
        else
        begin
          d2 := (ip.X - pt.X);
          d3 := (ipNext.X - pt.X);
          d := d2 * (ipNext.Y - pt.Y) - d3 * (ip.Y - pt.Y);
          if (d = 0) then Exit;
          if ((d > 0) = (ipNext.Y > ip.Y)) then val := 1 - val;
        end;
      end else
      begin
        if (ipNext.X > pt.X) then
        begin
          d2 := (ip.X - pt.X);
          d3 := (ipNext.X - pt.X);
          d := d2 * (ipNext.Y - pt.Y) - d3 * (ip.Y - pt.Y);
          if (d = 0) then Exit;
          if ((d > 0) = (ipNext.Y > ip.Y)) then val := 1 - val;
        end;
      end;
    end;
    ip := ipNext;
  end;

  case val of
    -1: result := pipOn;
     1: result := pipInside;
     else result := pipOutside;
  end;
end;
//---------------------------------------------------------------------------

function CrossProduct(const pt1, pt2, pt3: TPoint64): double;
var
  x1,x2,y1,y2: double;
begin
  x1 := pt2.X - pt1.X;
  y1 := pt2.Y - pt1.Y;
  x2 := pt3.X - pt2.X;
  y2 := pt3.Y - pt2.Y;
  result := (x1 * y2 - y1 * x2);
end;
//------------------------------------------------------------------------------
//------------------------------------------------------------------------------


end.

File: cpp/clipper_core.cpp
/*******************************************************************************
* Author    :  Angus Johnson                                                   *
* Version   :  10.0 (beta)                                                     *
* Date      :  18 March 2019                                                   *
* Website   :  http://www.angusj.com                                           *
* Copyright :  Angus Johnson 2010-2019                                         *
* Purpose   :  Core Clipper Library module                                     *
*              Contains structures and functions used throughout the library   *
* License   :  http://www.boost.org/LICENSE_1_0.txt                            *
*******************************************************************************/

#include "clipper_core.h"

namespace clipperlib {

//------------------------------------------------------------------------------
// Init templates  ...
//------------------------------------------------------------------------------

template struct Point<cInt>;
template struct Point<double>;
template struct Rect<cInt>;
template struct Rect<double>;
template struct Path<cInt>;
template struct Path<double>;
template struct Paths<cInt>;
template struct Paths<double>;
template struct PathsArray<cInt>;
template struct PathsArray<double>;

//------------------------------------------------------------------------------
// Point
//------------------------------------------------------------------------------

void PointI::Rotate(const PointD &center, double angle_rad) {
	double tmp_x = x - center.x;
	double tmp_y = y - center.y;
	double cos_a = cos(angle_rad);
	double sin_a = sin(angle_rad);

	x = (cInt)round(tmp_x * cos_a - tmp_y * sin_a + center.x);
	y = (cInt)round(tmp_x * sin_a - tmp_y * cos_a + center.y);
}
//------------------------------------------------------------------------------

void PointD::Rotate(const PointD &center, double angle_rad) {
	double tmp_x = x - center.x;
	double tmp_y = y - center.y;
	double cos_a = cos(angle_rad);
	double sin_a = sin(angle_rad);

	x = tmp_x * cos_a - tmp_y * sin_a + center.x;
	y = tmp_x * sin_a - tmp_y * cos_a + center.y;
}
//------------------------------------------------------------------------------

void PointI::Rotate(const PointD &center, double sin_a, double cos_a) {
	double tmp_x = x - center.x;
	double tmp_y = y - center.y;

	x = (cInt)round(tmp_x * cos_a - tmp_y * sin_a + center.x);
	y = (cInt)round(tmp_x * sin_a - tmp_y * cos_a + center.y);
}
//------------------------------------------------------------------------------

void PointD::Rotate(const PointD &center, double sin_a, double cos_a) {
	double tmp_x = x - center.x;
	double tmp_y = y - center.y;

	x = tmp_x * cos_a - tmp_y * sin_a + center.x;
	y = tmp_x * sin_a - tmp_y * cos_a + center.y;
}

//------------------------------------------------------------------------------
// Rect
//------------------------------------------------------------------------------

template <typename T>
void Rect<T>::Intersect(const Rect<T> &rect) {
	if (IsEmpty())
		return;
	else if (rect.IsEmpty()) {
		*this = Rect();
	} else {
		left = std::max(rect.left, left);
		right = std::min(rect.right, right);
		top = std::max(rect.top, top);
		bottom = std::min(rect.bottom, bottom);
		if (IsEmpty())
			*this = Rect();
	}
}
//------------------------------------------------------------------------------

void RectI::Rotate(double angle_rad) {
	PointD cp;
	PathD pts;

	cp.x = double(right + left) / 2;
	cp.y = double(bottom + top) / 2;

	pts.resize(4);
	pts[0] = PointD((double)left, (double)top);
	pts[1] = PointD((double)right, (double)top);
	pts[2] = PointD((double)right, (double)bottom);
	pts[3] = PointD((double)left, (double)bottom);
	pts.Rotate(cp, angle_rad);

	left = (cInt)floor(pts[0].x);
	top = (cInt)floor(pts[0].y);
	right = (cInt)ceil(pts[0].x);
	bottom = (cInt)ceil(pts[0].y);

	for (const auto &p : pts.data) {
		if (p.x < left) left = (cInt)floor(p.x);
		if (p.y < top) top = (cInt)floor(p.y);
		if (p.x > right) right = (cInt)ceil(p.x);
		if (p.y > bottom) bottom = (cInt)ceil(p.y);
	}
}
//------------------------------------------------------------------------------

void RectD::Rotate(double angle_rad) {
	PointD cp;
	PathD pts;

	cp.x = double(right + left) / 2;
	cp.y = double(bottom + top) / 2;

	pts.resize(4);
	pts[0] = PointD(left, top);
	pts[1] = PointD(right, top);
	pts[2] = PointD(right, bottom);
	pts[3] = PointD(left, bottom);

	pts.Rotate(cp, angle_rad);
	left = pts[0].x;
	top = pts[0].y;
	right = pts[0].x;
	bottom = pts[0].y;

	for (const auto &point : pts.data) {
		if (point.x < left) left = point.x;
		if (point.y < top) top = point.y;
		if (point.x > right) right = point.x;
		if (point.y > bottom) bottom = point.y;
	}
}
//------------------------------------------------------------------------------

template <typename T>
void Rect<T>::Union(const Rect<T> &rect) {
	if (rect.IsEmpty())
		return;
	else if (IsEmpty()) {
		*this = rect;
		return;
	}
	left = std::min(rect.left, left);
	right = std::max(rect.right, right);
	top = std::min(rect.top, top);
	bottom = std::max(rect.bottom, bottom);
}

//------------------------------------------------------------------------------
// Path
//------------------------------------------------------------------------------

PathI::Path(const PathI &other, double scale_x, double scale_y) {
	if (scale_x == 0) scale_x = 1;
	if (scale_y == 0) scale_y = 1;
	if (scale_x == 1 && scale_y == 1) {
		Append(other);
	} else {
		data.reserve(other.size());
		for (const auto &p : other.data)
			data.push_back(PointI((cInt)round(p.x * scale_x), (cInt)round(p.y * scale_y)));
	}
}
//------------------------------------------------------------------------------

PathI::Path(const PathD &other, double scale_x, double scale_y) {
	if (scale_x == 0) scale_x = 1;
	if (scale_y == 0) scale_y = 1;
	data.reserve(other.size());
	for (const auto &p : other.data)
		data.push_back(PointI((cInt)round(p.x * scale_x), (cInt)round(p.y * scale_y)));
}
//------------------------------------------------------------------------------

PathD::Path(const PathI &other, double scale_x, double scale_y) {
	if (scale_x == 0) scale_x = 1;
	if (scale_y == 0) scale_y = 1;
	data.reserve(other.size());
	for (const auto &p : other.data)
		data.push_back(PointD(p.x * scale_x, p.y * scale_y));
}
//------------------------------------------------------------------------------

PathD::Path(const PathD &other, double scale_x, double scale_y) {
	if (scale_x == 0) scale_x = 1;
	if (scale_y == 0) scale_y = 1;
	if (scale_x == 1 && scale_y == 1) {
		Append(other);
	} else {
		data.reserve(other.size());
		for (const auto &p : other.data)
			data.push_back(PointD(p.x * scale_x, p.y * scale_y));
	}
}
//------------------------------------------------------------------------------

PathI &PathI::operator=(const PathD &other) {
	data.clear();
	for (const auto &p : other.data)
		data.push_back(PointI((cInt)round(p.x), (cInt)round(p.y)));
	return *this;
}
//------------------------------------------------------------------------------

PathD &PathD::operator=(const PathI &other) {
	data.clear();
	for (const auto &p : other.data)
		data.push_back(PointD((double)p.x, (double)p.y));
	return *this;
}
//------------------------------------------------------------------------------

PathI &PathI::operator=(const PathI &other) {
	if (this != &other) {
		data.clear();
		Append(other);
	}
	return *this;
}
//------------------------------------------------------------------------------

PathD &PathD::operator=(const PathD &other) {
	if (this != &other) {
		data.clear();
		Append(other);
	}
	return *this;
}
//------------------------------------------------------------------------------

template <typename T>
void Path<T>::Append(const Path<T> &extra) {
	data.insert(end(data), begin(extra.data), end(extra.data));
}
//------------------------------------------------------------------------------

template <typename T>
double Path<T>::Area() const {
	double area = 0.0;
	auto len = data.size() - 1;
	if (len < 3) return area;
	auto j = len - 1;
	for (decltype(len) i = 0; i < len; ++i) {
		double d = (double)(data[j].x + data[i].x);
		area += d * (data[j].y - data[i].y);
		j = i;
	}
	return -area * 0.5;
}
//------------------------------------------------------------------------------

template <typename T>
Rect<T> Path<T>::Bounds() const {
	const T _MAX = std::numeric_limits<T>::max();
	const T _MIN = -_MAX;

	Rect<T> bounds(_MAX, _MAX, _MIN, _MIN);

	for (const auto &point : data) {
		if (point.x < bounds.left) bounds.left = point.x;
		if (point.x > bounds.right) bounds.right = point.x;
		if (point.y < bounds.top) bounds.top = point.y;
		if (point.y > bounds.bottom) bounds.bottom = point.y;
	}

	if (bounds.left >= bounds.right)
		return Rect<T>();
	else
		return bounds;
}
//------------------------------------------------------------------------------

template <typename T>
void Path<T>::Offset(T dx, T dy) {
	if (dx == 0 && dy == 0) return;
	for (auto &point : data) {
		point.x += dx;
		point.y += dy;
	}
}
//------------------------------------------------------------------------------

template <typename T>
bool Path<T>::Orientation() const {
	return Area() >= 0;
}
//------------------------------------------------------------------------------

template <typename T>
void Path<T>::Reverse() {
	std::reverse(begin(data), end(data));
}
//------------------------------------------------------------------------------

template <typename T>
void Path<T>::Rotate(const PointD &center, double angle_rad) {
	double cos_a = cos(angle_rad);
	double sin_a = sin(angle_rad);

	for (auto &point : data)
		point.Rotate(center, sin_a, cos_a);
}
//------------------------------------------------------------------------------

template <typename T>
void Path<T>::Scale(T sx, T sy) {
	if (sx == 0) sx = 1;
	if (sy == 0) sy = 1;
	if (sx == 1 && sy == 1) return;

	for (auto &point : data) {
		point.x *= sx;
		point.y *= sy;
	}
	StripDuplicates();
}
//------------------------------------------------------------------------------

template <typename T>
void Path<T>::StripDuplicates() {
	data.erase(unique(begin(data), end(data)), end(data));
}

//------------------------------------------------------------------------------
// Paths
//------------------------------------------------------------------------------

template <typename T>
Rect<T> Paths<T>::Bounds() const {
	const T _MAX = std::numeric_limits<T>::max();
	const T _MIN = -_MAX;

	Rect<T> bounds(_MAX, _MAX, _MIN, _MIN);

	for (const auto &path : data) {
		for (const auto &point : path.data) {
			if (point.x < bounds.left) bounds.left = point.x;
			if (point.x > bounds.right) bounds.right = point.x;
			if (point.y < bounds.top) bounds.top = point.y;
			if (point.y > bounds.bottom) bounds.bottom = point.y;
		}
	}

	if (bounds.left >= bounds.right)
		return Rect<T>();
	else
		return bounds;
}
//------------------------------------------------------------------------------

template <typename T>
void Paths<T>::Rotate(const PointD &center, double angle_rad) {
	double cos_a = cos(angle_rad);
	double sin_a = sin(angle_rad);

	for (auto &path : data)
		for (auto &point : path.data)
			point.Rotate(center, sin_a, cos_a);
}
//------------------------------------------------------------------------------

template <typename T>
void Paths<T>::Scale(T scale_x, T scale_y) {
	for (auto &path : data)
		path.Scale(scale_x, scale_y);
}
//------------------------------------------------------------------------------

template <typename T>
void Paths<T>::Offset(T dx, T dy) {
	if (dx == 0 && dy == 0) return;
	for (auto &path : data)
		for (auto &point : path.data) {
			point.x += dx;
			point.y += dy;
		}
}
//------------------------------------------------------------------------------

template <typename T>
void Paths<T>::Reverse() {
	for (auto &path : data)
		path.Reverse();
}
//------------------------------------------------------------------------------

template <typename T>
void Paths<T>::Append(const Paths<T> &extra) {
	data.insert(end(data), begin(extra.data), end(extra.data));
}

//------------------------------------------------------------------------------
// PathsArray
//------------------------------------------------------------------------------

template <typename T>
Rect<T> PathsArray<T>::Bounds() const {
	const T _MAX = std::numeric_limits<T>::max();
	const T _MIN = -_MAX;

	Rect<T> bounds(_MAX, _MAX, _MIN, _MIN);

	for (const auto &paths : data) {
		for (const auto &path : paths.data) {
			for (const auto &point : path.data) {
				if (point.x < bounds.left) bounds.left = point.x;
				if (point.x > bounds.right) bounds.right = point.x;
				if (point.y < bounds.top) bounds.top = point.y;
				if (point.y > bounds.bottom) bounds.bottom = point.y;
			}
		}
	}

	if (bounds.left >= bounds.right)
		return Rect<T>();
	else
		return bounds;
}

//------------------------------------------------------------------------------
// Miscellaneous
//------------------------------------------------------------------------------

PipResult PointInPolygon(const PointI &pt, const PathI &path) {
	int val = 0;
	auto cnt = path.size();
	double d, d2, d3;  // using doubles to avoid possible integer overflow
	PointI ip, ip_next;
	PipResult result = pipOnEdge;

	if (cnt < 3) {
		result = pipOutside;
		return result;
	}
	ip = path[0];
	for (decltype(cnt) i = 1; i < cnt; ++i) {
		if (i < cnt)
			ip_next = path[i];
		else
			ip_next = path[0];

		if (ip_next.y == pt.y) {
			if ((ip_next.x == pt.x) || ((ip.y == pt.y) && ((ip_next.x > pt.x) == (ip.x < pt.x)))) {
				return result;
			}
		}
		if ((ip.y < pt.y) != (ip_next.y < pt.y)) {
			if (ip.x >= pt.x) {
				if (ip_next.x > pt.x) {
					val = 1 - val;
				} else {
					d2 = (double)(ip.x - pt.x);
					d3 = (double)(ip_next.x - pt.x);
					d = d2 * (double)(ip_next.y - pt.y) - d3 * (double)(ip.y - pt.y);
					if (d == 0)
						return result;
					if ((d > 0) == (ip_next.y > ip.y))
						val = 1 - val;
				}
			} else {
				if (ip_next.x > pt.x) {
					d2 = (double)(ip.x - pt.x);
					d3 = (double)(ip_next.x - pt.x);
					d = d2 * (double)(ip_next.y - pt.y) - d3 * (double)(ip.y - pt.y);
					if (d == 0)
						return result;
					if ((d > 0) == (ip_next.y > ip.y))
						val = 1 - val;
				}
			}
		}
		ip = ip_next;
	}
	switch (val) {
		case -1: result = pipOnEdge; break;
		case 1: result = pipInside; break;
		default: result = pipOutside; break;
	}
	return result;
}
//------------------------------------------------------------------------------

double CrossProduct(const PointI &pt1, const PointI &pt2, const PointI &pt3) {
	double x1 = double(pt2.x - pt1.x);
	double y1 = double(pt2.y - pt1.y);
	double x2 = double(pt3.x - pt2.x);
	double y2 = double(pt3.y - pt2.y);
	return (x1 * y2 - y1 * x2);
}
//------------------------------------------------------------------------------
//------------------------------------------------------------------------------

}  // namespace clipperlib
